Track negated options in the option parser

eklitzke opened the pull request:
This change explicitly enable tracking negated options in the option parser. A negated option is one passed with a `-no` prefix. For example, `-nofoo` is the negated form of `-foo`. Negated options were originally added in the 0.6 release.

The change here allows code to explicitly distinguish between cases like `-nofoo` and `-foo=0`, which was not possible previously. The option parser does not have any changed semantics as a result of this change, and existing code will parse options just as it did before.

The motivation for this change is to provide a way to disable options that are otherwise not boolean options. For example, the `-debuglogfile` option is normally interpreted as a string, where the value is the log file name. With this change a user can pass in `-nodebuglogfile` and the code can see that it was explicitly negated, and use that to disable the log file.

This change originally split out from #12689.

MeshCollider commented:
Looks good, thanks for splitting it out. utACK https://github.com/bitcoin/bitcoin/pull/12713/commits/6ad2a13cfdd46899c28f21899e107ee7d4c07877

MarcoFalke commented:
utACK 6ad2a13cfdd46899c28f21899e107ee7d4c07877 (Also noting that `-a=no` sets a to true, now)

eklitzke commented:
Yeah I didn't want to add support for strings like "no" "disabled" etc. because I don't think there was ever an intention to support non-integer parameters in the first place. If we wanted to support that it's easy enough. You can also use `-noa`, both now and before.

donaloconnor commented:
utACK https://github.com/bitcoin/bitcoin/commit/6ad2a13cfdd46899c28f21899e107ee7d4c07877

jimpo commented:
ACK 6ad2a13cfdd46899c28f21899e107ee7d4c07877.

promag commented on src/util.cpp:
```diff
@@ -422,16 +422,14 @@ void ReleaseDirectoryLocks()
-    if (strValue.empty())
-        return true;
-    return (atoi(strValue) != 0);
+    // These values mean false, anything else (including empty string) is true.
```
Why not throw an error if it's a stupid value? IMHO, for instance, `-txindex=foo` should fail.

MarcoFalke replied:
Thinking more about this, with proper argument validation we might as well reject `false` as a value for a boolean argument. This was never supported, so it feels weird to special case it now.

eklitzke replied:
@promag Two reasons. First, such a change would not be backwards compatible (a minor point, see below). A better reason is that it enables the `-printtoconsole` change to have a way of disabling the debug log file without actually adding new command line arguments. In other words, I first test if `-debuglogfile` boolean parses as "false" (i.e. `-nodebuglogfile` or `-debuglogfile=0`), and if it's not false I re-parse it as a string. There are already a *lot* of possible flags to give Bitcoin, and if we can re-use existing functionality without adding new flags that seems better to me.

@MarcoFalke I actually agree, as accepting `false` opens the door for `disabled`, `untrue`, `nil`, `void`, `zip`, `null`, `non`, `nein`, ad infinitum.

My preferred semantics would be: `-nofoo` sets `foo` to false. Also `-foo=0` sets `foo` to false. Which is exactly what the *intention* of the existing code seems to be. I only handled the `false` flag as there is existing code committed in the Bitcoin repository that relies on it.

I started this endeavor just wanting to get bitcoind printing to stdout when run in interactive mode. Whatever semantics allow that to happen are fine with me.

ryanofsky replied:
> I first test if -debuglogfile boolean parses as "false" (i.e. -nodebuglogfile or -debuglogfile=0), and if it's not false I re-parse it as a string.

Feel free to discount this opinion, but if the option is `-debuglogfile=<filename>`, I would hope that `-debuglogfile=0` or `-debuglogfile=false` would just create log debug log files `0` or `false` respectively, instead of disabling logging.

Like João and Marco, I would also want ambiguous values passed as boolean options to trigger errors.

I guess I'm not so much a fan of option parsing trying to be very clever.

hkjn replied:
+1 to what @ryanofsky says; if backwards-compatability is not a constraint, doing the least clever thing seems most consistent and likely to lead to the most maintainable code, and long term least confusion for users.

(This PR seems like an improvement in readability of the argument parsing code, so any further simplification might be for another changeset.)

MarcoFalke commented:
Added "refactoring" label, since it doesn't change behavior

utACK e944ec998fbd232b05ec9471ae6f19d1d32e0719


Would you mind updating the OP, which ends up in the merge commit?

hkjn commented:
ACK 6ad2a13.

I'll note that we accept some pretty wacky stuff that we may (in separate PRs) want to change to exit with errors on:

```
$ bitcoind -noprinttoconsole       # false
$ bitcoind -printtoconsole=no      # true
$ bitcoind -printtoconsole=false   # false
$ bitcoind -noprinttoconsole=1     # false
$ bitcoind -noprinttoconsole=0     # true
$ bitcoind -noprinttoconsole=false # true
$ bitcoind -noprinttoconsole=yes   # false
```

jnewbery commented:
> since it doesn't change behavior

This is a change in behaviour. If I take the changes to the unit test, but not the changes to util.cpp, the unit tests fail. If I take the changes to util.cpp, but not the change to wallet_resendwallettransactions.py, then wallet_resendwallettransactions.py fails.

eklitzke commented:
These are all easy to implement. I can make `-debuglogfile=0` log to a file named `0`, I can make it not log at all, I can special case `-nodebuglogfile` and have that not create a file, I can parse `false` as false or parse it as true, I can make a backwards compatible change and I can make a backwards incompatible change.

I'm not really sure what to do, as this has gotten an unusual amount of debate for something that is frankly quite banal. I got three different suggestions for how to disable the flag in the original PR:

 * -nodebuglogfile
 * -debuglogfile=0
 * -debuglogfile=/dev/null

If there is not consensus on whether we should change this I will just drop this change altogether and add a brand new flag in the original PR (-disablelogfile).

ryanofsky commented:
> I'm not really sure what to do, as this has gotten an unusual amount of debate for something that is frankly quite banal. I got three different suggestions for how to disable the flag in the original PR:

From my reading of the feedback, I think if you go with `-nodebuglogfile` to disable logging, `-debuglogfile=<file>` to enable logging to a custom location, and `-debuglogfile` and `-debuglogfile=` to enable logging to the default location, you will have no objections. The ambivalence just came from zip/nein/0 stuff, which can be avoided by treating all nonempty string values the same.

hkjn commented:
ACK e944ec9.

This behavior as well as code seems slightly simpler than before, with more
tests.

I'll note we accept some pretty wacky combinations even after this change though,
that we may want to make errors in the future, at least for boolean options:

```
$ bitcoind -printtoconsole=0       # false
$ bitcoind -noprinttoconsole       # false
$ bitcoind -noprinttoconsole=no    # false
$ bitcoind -noprinttoconsole=false # false
$ bitcoind -noprinttoconsole=1     # false
$ bitcoind -printtoconsole=1       # true
$ bitcoind -printtoconsole=no      # true
$ bitcoind -printtoconsole=false   # true
$ bitcoind -noprinttoconsole=0     # true
```

eklitzke commented:
@ryanofsky I will put up another version of this diff with that change.

eklitzke commented:
Handle arguments like `-nofoo` more explicitly.

This change keeps track of whether an option was passed with the `-no` semantics. This way in #12689 I can do something like:

```c++
if (gArgs.IsArgInverted("-debuglogfile")) {
   ....
}
```

to see if the debug log file should be disabled.

I also removed some gratuitous boost and string copies.

@jnewbery @MarcoFalke  You guys figure out how you want the `-foo=false` case should be handled and let me know what you decide on.

eklitzke commented:
Updated for more consistent naming -- instead of "negated" and "inverted" I now always use the negated terminology.

MarcoFalke commented on src/util.h:
```diff
@@ -242,6 +245,15 @@ class ArgsManager
+     * @param strArg Argument to get (e.g. "-foo")
+     * @return true if the argument was passed negated
+     */
+    bool IsArgNegated(const std::string& strArg) const;
```
I don't think we should expose this in the code base. Historically, `-nofoo` always implied `-foo=0`, so use cases within the code  base should always compare with the string `"0"` to check if foo was unset.

Yes, I know that that would forbid use cases, such as `-myfile=0`, but imo that is acceptable and desirable to keep previous behavior.

ryanofsky replied:
> Historically, -nofoo always implied -foo=0, so use cases within the code base should always compare with the string "0" to check if foo was unset.

This is true for boolean options, and should still be true with this PR. I don't think it was ever historically true that specifying "0" or "false" for path type options would disable those options. At least this isn't the case currently for `-conf` `-debuglogfile` `-pid` `-rpccookiefile` `-wallet` or other path options. So whether or not special path strings are a good idea (I think they are a bad idea), I don't think there is a historic argument for having them.

MarcoFalke replied:
I wanted to raise the point that we specifically check for the string "0" in some non-bool arguments to see if they are disabled. E.g. `-connect=0` or `-debug=0`. So in case someone were to use `IsArgNegated` for those, it would break it.

Again, only bike-shedding here, so please ignore.

eklitzke replied:
I do kind of agree with @ryanofsky that not being able to create a log file named "0" (or "false") is weird. Tracking negation makes the arguments more strongly typed in some kind of weird sense. I also felt like it made my printtoconsole change a little easier to understand, because it removes the code that calls `GetBoolArg()` on a non-boolean field.

maflcko reviewed (commented):
re-utACK 6b95d6bd6e81030605d6b4ad42dd83cd3bc69bfd (Changes were to get rid of some boost usage and expose `IsArgNegated`, which I slightly object to)

ryanofsky reviewed (commented):
utACK 6b95d6bd6e81030605d6b4ad42dd83cd3bc69bfd.

It seems to me though that this is doing two different things that don't need to be tied to together:

1) Introducing a nice `IsArgNegated` API that provides a consistent way of disabling non-bool options, especially features that read or write to files.
2) Interpreting `-foo=bar` and similar strings as false instead of true.

Second change seems fine, but IMO would probably be better off in a separate PR. It could also be tweaked to accept more false strings, and should definitely be documented.

ryanofsky commented on src/test/util_tests.cpp:
```diff
@@ -223,6 +217,54 @@ BOOST_AUTO_TEST_CASE(util_ParseParameters)
     BOOST_CHECK(testArgs.GetArgs("-ccc").size() == 2);
 }
 
+BOOST_AUTO_TEST_CASE(util_GetBoolArg)
```
Great tests!

hkjn commented:
Re-ACK 6b95d6b.

+1 on documenting the change in behavior for flags in release notes.

eklitzke commented:
Last night I had the same realization that @ryanofsky pointed out here, which is that the change to `InterpretBoolArg()` is no longer necessary at all, as my printtoconsole change now uses the new `IsNegatedArg()` method (also originally suggested by Russ). I am just going to revert all of the unnecessary changes, and then this change won't change any backwards or forwards compatibility expectations. The code will still use `atoi()` as it does now. Users in an alternate universe who have `txindex=-0` in their bitcoin.conf won't have to worry about what happens if they upgrade to Bitcoin 0.17. I will add a comment in this area of the code summarizing the discussion here for the next person who wants to wade into this.

I do want to point out:

 * It's pretty clear the the original authors of this code didn't think about these edge cases, the existing tests don't check any of them, and this code makes no sense at all when you start thinking about the cases we've talked about here.
 * The return value of `atoi()` is undefined if the string it's given represents a value outside the range of `INT_MIN` to `INT_MAX`. So if we're going to worry about edge cases, it's not just that the existing code has strange semantics, it has undefined behavior. Almost all real-world computers today are [LLP64 (Windows) or LP64 (Unix)](https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models). This means that `txindex=2147483647` is well-defined on most people's computers, but `txindex=2147483648` is not. In my opinion this is a bug, and fixing a bug of this nature requires changing backwards compatibility expectations (for some definition of backwards compatibility).
 * In almost any other part of the code, if we found out that Bitcoin behaved in an unexpected (or undefined) way when given pathological input we would fix it even though that's not strictly backwards compatible. My opinion (and it is an opinion only) is that changing how the option parser interprets data that it didn't expect to see is the same as changing other kinds of code that didn't properly validate input. I don't think there would be any hesitation to change code in other areas if we found out it didn't properly validate input values or check return codes, and to me this is the same variety of change.

eklitzke committed:
Add additional tests for GetBoolArg()

This is meant to be an intermediate commit to prove that the next does not
introduce any changes in the semantics of boolean option parsing.

eklitzke committed:
Track negated arguments in the argument paser.

This commit adds tracking for negated arguments. This change will be used in a
future commit that allows disabling the debug.log file using -nodebuglogfile.

eklitzke commented:
This branch has two commits:

 * The first is just a change to test_util.cpp with the new tests I added.
 * The second change adds the negated argument tracking, and more tests.

The reason this is two commits is to prove that the semantics of boolean option parsing is unchanged in this branch. I have tested both commits locally and both pass the test suite.

There are no longer any changes to `InterpretBool()`, other than comments summarizing the discussion here.

Rebasing and splitting out these commits is a lot of work. Please do not request additional changes unless there are any serious errors here.

ryanofsky reviewed (commented):
utACK ec1e226e33713366410e7d19a287be521169cd5f. The only change since my last review was reverting the `InterpretBool` diff and adding comments. Now this PR no longer changes behavior, only adding tests, comments, and a new `ArgsManager::InterpretNegatedOption()` method that makes it straightforward to implement options that can be disabled, like `-debuglogfile=<path>` / `-nodebuglogfile`.

> Rebasing and splitting out these commits is a lot of work. Please do not request additional changes unless there are any serious errors here.

Thanks for your work. I usually only suggest changes, not request them. You should always feel free to ignore any suggestions that come from me, or disagree with them, or respond with boilerplate like "this seems like a good idea but [would delay other work I'm doing / could be implemented later / does not seem worth the required effort right now]."

jnewbery commented:
ugh. Why? The previous branch fixed undesirable behaviour and already had 4 ACKs.

MarcoFalke commented:
re-utACK ec1e226e33713366410e7d19a287be521169cd5f

Needs simple rebase, it seems.

ryanofsky commented:
> ugh. Why? The previous branch fixed undesirable behaviour and already had 4 ACKs.

To be fair, only 6 lines of code changed between the current and previous versions of this PR. 
3 lines were reverted in InterpretBool, 1 line was reverted in wallet_resendwallettransactions.py, and 2 GetBoolArg checks in the test changed from true to false.

I don't think the previous behavior change was an unambiguous improvment either, and it wasn't documented or directly related to the feature introduced here.

jnewbery commented:
Fine. Re-utACK ec1e226e33713366410e7d19a287be521169cd5f.

Agree that documentation should be added when the `InterpretBool()` behaviour is changed.

promag commented on src/util.cpp:
```diff
@@ -419,21 +417,54 @@ void ReleaseDirectoryLocks()
+ * options that are not normally boolean (e.g. using -nodebuglogfile to request
+ * that debug log output is not sent to any file at all).
+ */
+void ArgsManager::InterpretNegatedOption(std::string& key, std::string& val)
```
Missing `const` on both?

promag commented on src/util.cpp:
```diff
@@ -442,34 +473,34 @@ void ArgsManager::ParseParameters(int argc, const char* const argv[])
-            str = str.substr(0, is_index);
+    m_negated_args.clear();
+
+    for (int i = 1; i < argc; i++) {
```
nit, since you are touching this, `++i`.

hkjn commented:
ACK ec1e226.

Here's the current interpretation of flags that I showed in my earlier comment (https://github.com/bitcoin/bitcoin/pull/12713#issuecomment-374563792), noting the differences from the earlier reviewed commit 6b95d6b:

```
$ bitcoind -printtoconsole=0       # false 
$ bitcoind -noprinttoconsole       # false 
$ bitcoind -noprinttoconsole=1     # false 
$ bitcoind -printtoconsole=no      # false (earlier true) 
$ bitcoind -printtoconsole=false   # false (earlier true) 
$ bitcoind -printtoconsole=1       # true 
$ bitcoind -noprinttoconsole=no    # true (earlier false) 
$ bitcoind -noprinttoconsole=0     # true 
$ bitcoind -noprinttoconsole=false # true (earlier false)
```

MarcoFalke commented:
@eklitzke Mind doing the rebase?

ryanofsky commented:
@eklitzke, you may want to note in the PR description that this PR doesn't change current behavior, except to now log warnings if double-negative -nofoo=0 options are encountered. The changes that this PR does make are:

- Adding a convenient way to check for disabled, non-boolean option pairs like `-debuglogfile=<path>`, `-nodebuglogfile`.
- Cleaning up code and variable names
- Adding lots of comments.

jnewbery commented:
Also:

- removing a boost algorithm #include

eklitzke commented:
Rebased, tests passed locally for me. Travis has been kind of flaky recently so hopefully it gives me a green build. I've also updated the PR description.

hkjn commented:
Re-utACK f7683cb.

promag commented:
utACK f7683cb.

MarcoFalke commented:
utACK f7683cba7b070b722a2e0641f4d1516112392ed6
